src/kernel/lsm/ss_table/loader.rs

```diff
@@ -0,0 +1,278 @@
+use crate::kernel::lsm::ss_table::footer::{Footer, TABLE_FOOTER_SIZE};
+use crate::kernel::lsm::ss_table::meta::SSTableMeta;
+use crate::kernel::lsm::storage::Config;
+use crate::kernel::utils::lru_cache::ShardingLruCache;
+use crate::kernel::Result;
+
+#[derive(Clone)]
+pub(crate) struct SSTableLoader {
+    inner: Arc<ShardingLruCache<i64, SSTable>>,
+    factory: Arc<IoFactory>,
+    config: Config,
+    wal: LogLoader
+}
+
+impl SSTableLoader {
+    pub(crate) fn new(config: Config, factory: Arc<IoFactory>, wal: LogLoader) -> Result<Self> {
+        let inner = Arc::new(
+            ShardingLruCache::new(
+                config.table_cache_size,
+                16,
+                RandomState::default()
+            )?
+        );
+        Ok(SSTableLoader { inner, factory, config, wal })
+    }
+
+    fn _create(
+        &self,
+        gen: i64,
+        vec_data: Vec<KeyValue>,
+        level: usize
+    ) -> Result<SSTable> {
+        fn io_type_by_level(level: usize) -> IoType {
+            if LEVEL_0 == level {
+                IoType::Mem
+            } else {
+                IoType::Direct
+            }
+        }
+
+        let io_type = io_type_by_level(level);
+        let io_factory = &self.factory;
+        let config = &self.config;
+        let len = vec_data.len();
+        let data_restart_interval = config.data_restart_interval;
+        let index_restart_interval = config.index_restart_interval;
+        let mut filter = GrowableBloom::new(config.desired_error_prob, len);
+
+        let mut builder = BlockBuilder::new(
+            BlockOptions::from(config)
+                .compress_type(CompressType::LZ4)
+                .data_restart_interval(data_restart_interval)
+                .index_restart_interval(index_restart_interval)
+        );
+        for data in vec_data {
+            let (key, value) = data;
+            let _ = filter.insert(&key);
+            builder.add((key, Value::from(value)));
+        }
+        let meta = MetaBlock {
+            filter,
+            len,
+            index_restart_interval,
+            data_restart_interval,
+        };
+
+        let (data_bytes, index_bytes) = builder.build()?;
+        let meta_bytes = bincode::serialize(&meta)?;
+        let footer = Footer {
+            level: level as u8,
+            index_offset: data_bytes.len() as u32,
+            index_len: index_bytes.len() as u32,
+            meta_offset: (data_bytes.len() + index_bytes.len()) as u32,
+            meta_len: meta_bytes.len() as u32,
+            size_of_disk: (data_bytes.len() + index_bytes.len() + meta_bytes.len() + TABLE_FOOTER_SIZE) as u32,
+        };
+        let mut writer = io_factory.writer(gen, io_type)?;
+        writer.write_all(
+            data_bytes.into_iter()
+                .chain(index_bytes)
+                .chain(meta_bytes)
+                .chain(bincode::serialize(&footer)?)
```

Comment:
sstable add footer, I have seen it before in leveldb.

Reply:
> sstable add footer, I have seen it before in leveldb.

reference from LevelDB